File: editor/src/de/project/ice/editor/editors/NumberEditor.java
package de.project.ice.editor.editors;

import com.kotcrab.vis.ui.widget.NumberSelector;
import com.kotcrab.vis.ui.widget.VisTextField;

import java.text.ParseException;

public abstract class NumberEditor<T extends Number> extends ValueEditor<T> {
    private VisTextField valueField;

    @Override
    protected void updateValue() {
        valueField.setText(String.valueOf(value));
    }

    @Override
    protected void createUi() {
        valueField = new VisTextField(String.valueOf(value));
        add(valueField);
    }

    @Override
    public void act(float delta) {
        try {
            value = parseValue();
        } catch (NumberFormatException ignore) {}

        super.act(delta);
    }

    protected T parseValue() {
        try {
            return parse(valueField.getText());
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return null;
    }

    protected abstract T parse(String value) throws ParseException;

    public static class FloatEditor extends NumberEditor<Float> {
        @Override
        protected Float parse(String value) throws ParseException {
            return Float.parseFloat(value);
        }
    }

    public static class DoubleEditor extends NumberEditor<Double> {
        @Override
        protected Double parse(String value) throws ParseException {
            return Double.parseDouble(value);
        }
    }

    public static class LongEditor extends NumberEditor<Long> {

        @Override
        protected Long parse(String value) throws ParseException {
            return Long.parseLong(value);
        }
    }

    public static class IntegerEditor extends NumberEditor<Integer> {
        private NumberSelector numberSelector;

        @Override
        protected void createUi() {
            numberSelector = new NumberSelector(null, value, 0, Integer.MAX_VALUE);
            add(numberSelector);
        }

        @Override
        protected Integer parseValue() {
            return numberSelector.getValue();
        }

        @Override
        protected void updateValue() {
            numberSelector.setValue(value);
        }

        @Override
        protected Integer parse(String value) throws ParseException {
            return Integer.parseInt(value);
        }
    }

    public static class ShortEditor extends NumberEditor<Short> {
        @Override
        protected Short parse(String value) throws ParseException {
            return Short.parseShort(value);
        }
    }
}

File: editor/src/de/project/ice/editor/editors/NumberEditor.kt
package de.project.ice.editor.editors

import java.text.ParseException

abstract class NumberEditor<T : Number> : ValueEditor<T>() {
    private var valueField: VisTextField? = null

    override fun updateValue() {
        valueField!!.text = value.toString()
    }

    override fun createUi() {
        valueField = VisTextField(value.toString())
        add<VisTextField>(valueField)
    }

    override fun act(delta: Float) {
        try {
            value = parseValue()
        } catch (ignore: NumberFormatException) {
        }

        super.act(delta)
    }

    protected open fun parseValue(): T? {
        try {
            return parse(valueField!!.text)
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return null
    }

    @Throws(ParseException::class)
    protected abstract fun parse(value: String): T

    class FloatEditor : NumberEditor<Float>() {
        @Throws(ParseException::class)
        override fun parse(value: String): Float {
            return java.lang.Float.parseFloat(value)
        }
    }

    class DoubleEditor : NumberEditor<Double>() {
        @Throws(ParseException::class)
        override fun parse(value: String): Double {
            return java.lang.Double.parseDouble(value)
        }
    }

    class LongEditor : NumberEditor<Long>() {

        @Throws(ParseException::class)
        override fun parse(value: String): Long {
            return java.lang.Long.parseLong(value)
        }
    }

    class IntegerEditor : NumberEditor<Int>() {
        private var numberSelector: NumberSelector? = null

        override fun createUi() {
            numberSelector = NumberSelector(null, value!!.toFloat(), 0f, Integer.MAX_VALUE.toFloat(), 1.0f, 0)
            add<NumberSelector>(numberSelector)
        }

        override fun parseValue(): Int? {
            return numberSelector!!.value.toInt()
        }

        override fun updateValue() {
            numberSelector!!.value = value!!.toFloat()
        }

        @Throws(ParseException::class)
        override fun parse(value: String): Int {
            return Integer.parseInt(value)
        }
    }

    class ShortEditor : NumberEditor<Short>() {
        @Throws(ParseException::class)
        override fun parse(value: String): Short {
            return java.lang.Short.parseShort(value)
        }
    }
}
